File: codes/python/547.省份数量.py
class UnionFind:
	def __init__(self, n):
		self.parent = [i for i in range(n)]
		self.size = [1 for _ in range(n)]
		self.count = n

	def find(self, p):
		while p != self.parent[p]:
			self.parent[p] = self.parent[self.parent[p]]
			p = self.parent[p]
		return p

	def union(self, p, q):
		rootP = self.find(p)
		rootQ = self.find(q)
		if rootP == rootQ:
			return
		if self.size[rootP] > self.size[rootQ]:
			self.parent[rootQ] = rootP
			self.size[rootP] += self.size[rootQ]
		else:
			self.parent[rootP] = rootQ
			self.size[rootQ] += self.size[rootP]
		self.count -= 1


class Solution:
	def findCircleNum(self, isConnected: List[List[int]]) -> int:
		uf = UnionFind(len(isConnected))
		for i in range(len(isConnected)):
			for j in range(i + 1, len(isConnected)):
				if isConnected[i][j] == 1:
					uf.union(i, j)
		return uf.count
	
	

File: codes/java/547.省份数量.java

class UnionFind {
	private int[] parent;
	private int count;
	private int[] size;

	public UnionFind(int n) {
		parent = new int[n];
		size = new int[n];
		count = n;
		for (int i = 0; i < n; i++) {
			parent[i] = i;
			size[i] = 1;
		}
	}

	public int find(int p) {
		while (p != parent[p]) {
			parent[p] = parent[parent[p]];
			p = parent[p];
		}
		return p;
	}

	public void union(int p, int q) {
		int rootP = find(p);
		int rootQ = find(q);
		if (rootP == rootQ) {
			return;
		}
		if (size[rootP] > size[rootQ]) {
			parent[rootQ] = rootP;
			size[rootP] += size[rootQ];
		} else {
			parent[rootP] = rootQ;
			size[rootQ] += size[rootP];
		}
		count--;
	}

	public int count() {
		return count;
	}
}

class Solution {
	public int findCircleNum(int[][] isConnected) {
		UnionFind uf = new UnionFind(isConnected.length);
		for (int i = 0; i < isConnected.length; i++) {
			for (int j = i + 1; j < isConnected.length; j++) {
				if (isConnected[i][j] == 1) {
					uf.union(i, j);
				}
			}
		}
		return uf.count();
	}
}